src/cpp/logger.cpp

```diff
@@ -0,0 +1,252 @@
+    std::string requestedLogfile(_logfile);
+    bool missingLogFile = _logfile.length() == 0;
+
+    // Make sure it can be written to
+    bool badLogFile = false;
+    if (_logfile.length() > 0) {
+        std::ofstream logstream;
+        logstream.open(_logfile, std::ios_base::app);
+        badLogFile = !logstream.good();
+
+        // Switch to STDOUT if we can't write to the specified file
+        if (badLogFile)
+            _logfile = "";
+    }
+
+    // Get the logging level
+    std::string level;
+    get_config_string(level, "SR_LOG_LEVEL", "", true);
+    bool missingLogLevel = level.length() == 0;
+    bool badLogLevel = false;
+    if (level.length() > 0) {
+        if (level.compare("QUIET") == 0)
+            _log_level = LLQuiet;
+        else if (level.compare("INFO") == 0)
+            _log_level = LLInfo;
+        else if (level.compare("DEBUG") == 0)
+            _log_level = LLDebug;
+        else if (level.compare("DEVELOPER") == 0)
+            _log_level = LLDeveloper;
+        else {
+            // Don't recognize the requested level; use default
+            badLogLevel = true;
+            _log_level = LLInfo;
+        }
+    }
+    else {
+        _log_level = LLInfo;
+    }
+
+    // Now that everything is configured, issue warning and
+    // error messages
+    if (missingLogFile) {
+        log_warning(
+            LLInfo,
+            "Environment variable SS_LOG_FILE is not set. "
+            "Defaulting to stdout"
+        );
+    }
+    if (missingLogLevel) {
+        log_warning(
+            LLInfo,
+            "Environment variable SS_LOG_LEVEL is not set. "
+            "Defaulting to INFO"
+        );
+    }
+    if (badLogFile) {
+        throw SRRuntimeException(
+            "Cannot write to file: " + requestedLogfile);
+    }
+    if (badLogLevel) {
+        throw SRRuntimeException(
+            "Unrecognized logging level: " + level);
+    }
+
+    // Done
+    _initialized = true;
+}
+
+/*!
+*   \brief Conditionally log data if the logging level is high enough
+*   \param level Minimum logging level for data to be logged
+*   \param data Text of data to be logged
+*/
+void Logger::log_data(SRLoggingLevel level, const std::string& data)
+{
+    // If we're not initialized, configure logging as an anonymous
+    // client. This can happen if a caller invokes a Dataset API point
+    // without initializing a Client object
+    if (!_initialized)
+        configure_logging("anonymous");
+
+    // Silently ignore logging more verbose than requested
+    if (level > _log_level)
+        return;
+
+    // Get the current timestamp
+    auto t = std::time(NULL);
+    auto tm = *std::localtime(&t);
+    auto timestamp = std::put_time(&tm, "%H-%M-%S");
+
+    // write the log data
+    // (There must be a cleaner way to write this!)
+    if (_logfile.length() > 0) {
+        std::ofstream logstream;
+        logstream.open(_logfile, std::ios_base::app);
```

Comment:
We can't keep the file open because another writer might need to open the file to write it and I think that triggers a filesystem lock.

You're right that I should be checking good() here and I will add that

Reply:
Added check for good(). On failure, roll over logging to console